File: process_24dsf.c
/*
 * Process 24DSF.txt - Ultra lean C processor.
 * Only accumulates 384 doubles (4 combos x 24 months x 4 metrics).
 * Uses a fixed-size hash table for entity->sector mapping.
 */
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <ctype.h>

/* 4 combos x 24 months x 4 metrics = 384 */
#define N_COMBOS 4
#define N_MONTHS 24
#define N_METRICS 4
#define BLOCK_START 18
#define BLOCK_SIZE 15

/* Simple hash table for entity code -> sector (0=fin, 1=nofin) */
#define HT_SIZE 2048
typedef struct { int code; int sector; } ht_entry;
static ht_entry ht[HT_SIZE];
static int ht_used = 0;

static int classify_entity(int code) {
    if (code < 1000) return 0;
    if (code >= 44000 && code <= 45999) return 0;
    if (code >= 65000 && code <= 65999) return 0;
    return 1;
}

static void ht_insert(int code, int sector) {
    unsigned h = (unsigned)code % HT_SIZE;
    while (ht[h].code != -1) {
        if (ht[h].code == code) { ht[h].sector = sector; return; }
        h = (h + 1) % HT_SIZE;
    }
    ht[h].code = code;
    ht[h].sector = sector;
    ht_used++;
}

static int ht_lookup(int code) {
    unsigned h = (unsigned)code % HT_SIZE;
    while (ht[h].code != -1) {
        if (ht[h].code == code) return ht[h].sector;
        h = (h + 1) % HT_SIZE;
    }
    return -1;
}

/* accumulators: [combo][month][metric] where metric: 0=tc, 1=ic, 2=tr, 3=ir */
static double acc[N_COMBOS][N_MONTHS][N_METRICS];

static double parse_amount(const char *buf, int len) {
    /* buf is raw bytes from file, len chars. Replace ',' with '.', parse as double */
    char tmp[20];
    int j = 0;
    int has_content = 0;
    for (int i = 0; i < len && j < 19; i++) {
        char c = buf[i];
        if (c == ' ') continue;
        if (c == ',') { tmp[j++] = '.'; has_content = 1; }
        else if (isdigit(c) || c == '-') { tmp[j++] = c; if (c != '0') has_content = 1; }
    }
    if (!has_content || j == 0) return 0.0;
    tmp[j] = '\0';
    return atof(tmp);
}

static int is_irregular(const char *sit, int len) {
    /* Situations 3, 4, 5, 11 */
    /* sit is 2 chars, possibly space-padded */
    char s[4];
    int j = 0;
    for (int i = 0; i < len && j < 3; i++) {
        if (sit[i] != ' ') s[j++] = sit[i];
    }
    s[j] = '\0';
    if (j == 0) return 0;
    if (strcmp(s, "3") == 0 || strcmp(s, "4") == 0 || strcmp(s, "5") == 0 || strcmp(s, "11") == 0) return 1;
    return 0;
}

int main(int argc, char **argv) {
    const char *maeent_path = "/sessions/practical-upbeat-ride/mnt/202601DEUDORES/Maeent.txt";
    const char *input_path = "/sessions/practical-upbeat-ride/mnt/202601DEUDORES/DATAHISTORICA/24DSF.txt";
    const char *output_path = "/sessions/practical-upbeat-ride/data_series.json";

    /* Init hash table */
    for (int i = 0; i < HT_SIZE; i++) ht[i].code = -1;

    /* Load Maeent */
    FILE *fm = fopen(maeent_path, "r");
    if (!fm) { fprintf(stderr, "Cannot open Maeent\n"); return 1; }
    char line[1024];
    while (fgets(line, sizeof(line), fm)) {
        if (strlen(line) < 6) continue;
        char codestr[6];
        strncpy(codestr, line, 5);
        codestr[5] = '\0';
        int code = atoi(codestr);
        ht_insert(code, classify_entity(code));
    }
    fclose(fm);
    fprintf(stderr, "Loaded %d entities\n", ht_used);

    /* Init accumulators */
    memset(acc, 0, sizeof(acc));

    /* Process 24DSF */
    FILE *fi = fopen(input_path, "r");
    if (!fi) { fprintf(stderr, "Cannot open 24DSF\n"); return 1; }

    /* Use a large buffer for fgets */
    char buf[512];
    long long n = 0;
    int min_len = BLOCK_START + BLOCK_SIZE;

    while (fgets(buf, sizeof(buf), fi)) {
        n++;
        if ((n % 10000000) == 0) fprintf(stderr, "  %lld lines...\n", n);

        int blen = strlen(buf);
        if (blen < min_len) continue;

        /* CUIT prefix at pos 7-8 */
        char cp0 = buf[7], cp1 = buf[8];
        int tipo;
        if ((cp0 == '2' && (cp1 == '0' || cp1 == '3' || cp1 == '4' || cp1 == '7'))) {
            tipo = 0; /* familia */
        } else if ((cp0 == '3' && (cp1 == '0' || cp1 == '3' || cp1 == '4'))) {
            tipo = 1; /* empresa */
        } else {
            continue;
        }

        /* Entity code: pos 0-4 */
        char ecstr[6];
        strncpy(ecstr, buf, 5);
        ecstr[5] = '\0';
        int ec = atoi(ecstr);

        int sector = ht_lookup(ec);
        if (sector < 0) sector = classify_entity(ec);

        int combo = sector * 2 + tipo; /* 0=fin_fam, 1=fin_emp, 2=nofin_fam, 3=nofin_emp */

        /* Process 24 monthly blocks */
        for (int mi = 0; mi < N_MONTHS; mi++) {
            int off = BLOCK_START + mi * BLOCK_SIZE;
            if (off + 14 > blen) break;

            /* Situation: 2 chars at off */
            char sit[3] = { buf[off], buf[off+1], '\0' };

            /* Skip empty/zero situation */
            int sit_trimmed = 0;
            for (int i = 0; i < 2; i++) if (sit[i] != ' ' && sit[i] != '0') sit_trimmed = 1;
            if (!sit_trimmed) continue;

            /* Parse amount: 12 chars at off+2 */
            double amt = parse_amount(buf + off + 2, 12);

            acc[combo][mi][0] += amt;  /* tc */
            acc[combo][mi][2] += 1.0;  /* tr */

            if (is_irregular(sit, 2)) {
                acc[combo][mi][1] += amt;  /* ic */
                acc[combo][mi][3] += 1.0;  /* ir */
            }
        }
    }
    fclose(fi);
    fprintf(stderr, "Done: %lld lines\n", n);

    /* Generate output JSON */
    FILE *fo = fopen(output_path, "w");
    if (!fo) { fprintf(stderr, "Cannot open output\n"); return 1; }

    /* Month labels */
    const char *mnames[] = {"","Ene","Feb","Mar","Abr","May","Jun","Jul","Ago","Sep","Oct","Nov","Dic"};
    char months[24][8];
    char labels[24][16];
    int y = 2026, m = 1;
    for (int i = 0; i < 24; i++) {
        sprintf(months[i], "%04d-%02d", y, m);
        sprintf(labels[i], "%s %02d", mnames[m], y % 100);
        m--;
        if (m == 0) { m = 12; y--; }
    }

    /* Write JSON */
    fprintf(fo, "{\"months\":[");
    for (int i = 23; i >= 0; i--) { fprintf(fo, "\"%s\"%s", months[i], i > 0 ? "," : ""); }
    fprintf(fo, "],\"month_labels\":[");
    for (int i = 23; i >= 0; i--) { fprintf(fo, "\"%s\"%s", labels[i], i > 0 ? "," : ""); }
    fprintf(fo, "],\"series\":[");

    const char *keys[] = {"fam_fin", "emp_fin", "fam_nofin", "emp_nofin"};

    for (int i = 23; i >= 0; i--) {
        fprintf(fo, "{\"periodo\":\"%s\",\"label\":\"%s\"", months[i], labels[i]);
        for (int c = 0; c < 4; c++) {
            double tc = acc[c][i][0], ic = acc[c][i][1];
            double tr = acc[c][i][2], ir = acc[c][i][3];
            double pa = (tc > 0) ? (ic / tc * 100.0) : 0;
            double pq = (tr > 0) ? (ir / tr * 100.0) : 0;
            fprintf(fo, ",\"%s_amt\":%.2f,\"%s_qty\":%.2f", keys[c], pa, keys[c], pq);
            fprintf(fo, ",\"%s_tc\":%.1f,\"%s_ic\":%.1f", keys[c], tc, keys[c], ic);
            fprintf(fo, ",\"%s_tr\":%d,\"%s_ir\":%d", keys[c], (int)tr, keys[c], (int)ir);
        }
        fprintf(fo, "}%s", i > 0 ? "," : "");
    }
    fprintf(fo, "]}");
    fclose(fo);

    /* Print summary */
    fprintf(stderr, "\n=== SERIE HISTORICA (familias, por monto) ===\n");
    for (int i = 23; i >= 0; i--) {
        double tc_ff = acc[0][i][0], ic_ff = acc[0][i][1];
        double tc_nf = acc[2][i][0], ic_nf = acc[2][i][1];
        double pf = (tc_ff > 0) ? ic_ff/tc_ff*100 : 0;
        double pn = (tc_nf > 0) ? ic_nf/tc_nf*100 : 0;
        fprintf(stderr, "  %s: Fin=%6.2f%%  NoFin=%6.2f%%\n", labels[i], pf, pn);
    }

    fprintf(stderr, "\nSaved to %s\n", output_path);
    return 0;
}

File: scripts_nodejs/process_24dsf.js
#!/usr/bin/env node
/*
 * Procesa 24DSF.txt (23GB) y Maeent.txt
 * Equivalente Node.js de process_24dsf_v2.c
 * Genera:
 *   data_series.json        - serie agregada 24 meses por sector
 *   data_entity_series.json - serie por entidad por mes
 */
'use strict';

const fs = require('fs');
const path = require('path');

const INPUT_DIR = process.argv[2] || 'C:/Users/Igor/Downloads/Muni Pinamar/202601DEUDORES/24DSF202603';
const MAEENT_DIR = process.argv[3] || 'C:/Users/Igor/Downloads/Muni Pinamar/202601DEUDORES/202603DEUDORES';
const OUTPUT_DIR = process.argv[4] || 'C:/Users/Igor/Downloads/irregularidadprestamosARG';
const LATEST_YEAR = parseInt(process.argv[5] || '2026', 10);
const LATEST_MONTH = parseInt(process.argv[6] || '3', 10);  // March 2026 = block 0

const INPUT_FILE = path.join(INPUT_DIR, '24DSF.txt');
const MAEENT_FILE = path.join(MAEENT_DIR, 'Maeent.txt');

const N_MONTHS = 24;
const BLOCK_START = 18;
const BLOCK_SIZE = 15;

const IRREG_SITS = new Set(['3','4','5','11']);

function classifyEntity(code) {
  const c = parseInt(code, 10);
  if (c < 1000) return 0; // financiero
  if (c >= 44000 && c <= 45999) return 0;
  if (c >= 65000 && c <= 65999) return 0;
  return 1; // no_financiero
}

function loadEntities() {
  const entities = {}; // code (zero-padded) -> {sector, name}
  const data = fs.readFileSync(MAEENT_FILE, 'latin1');
  for (const line of data.split(/\r?\n/)) {
    if (line.length < 6) continue;
    const code = line.substring(0, 5).trim().padStart(5, '0');
    const name = line.substring(5).trim();
    entities[code] = { sector: classifyEntity(code), name };
  }
  return entities;
}

function parseAmount(buf) {
  // buf is a string of digits/spaces/comma. Replace comma with dot, parse.
  let tmp = '';
  let hasContent = false;
  for (let i = 0; i < buf.length; i++) {
    const c = buf[i];
    if (c === ' ') continue;
    if (c === ',') { tmp += '.'; hasContent = true; }
    else if ((c >= '0' && c <= '9') || c === '-') { tmp += c; if (c !== '0') hasContent = true; }
  }
  if (!hasContent || tmp.length === 0) return 0;
  const v = parseFloat(tmp);
  return Number.isFinite(v) ? v : 0;
}

function isIrregular(sit) {
  const trimmed = sit.replace(/\s/g, '');
  return IRREG_SITS.has(trimmed);
}

// MonthData: tc, ic, tr, ir as flat arrays for performance
// agg_data[combo][month] -> {tc,ic,tr,ir}
// 4 combos: 0=fin_fam, 1=fin_emp, 2=nofin_fam, 3=nofin_emp
const N_COMBOS = 4;
const aggTC = new Float64Array(N_COMBOS * N_MONTHS);
const aggIC = new Float64Array(N_COMBOS * N_MONTHS);
const aggTR = new Int32Array(N_COMBOS * N_MONTHS);
const aggIR = new Int32Array(N_COMBOS * N_MONTHS);

// Per-entity per-tipo per-month
// key: ec (5-digit) + tipo (familia/empresa)
// value: array of N_MONTHS objects
const entData = new Map(); // key -> { sector, name, tipo, monthsTC, monthsIC, monthsTR, monthsIR }

function getOrAddEntKey(entities, ec, tipo) {
  const key = `${ec}_${tipo === 0 ? 'familia' : 'empresa'}`;
  let e = entData.get(key);
  if (!e) {
    const meta = entities[ec] || { sector: classifyEntity(ec), name: '' };
    e = {
      code: ec,
      name: meta.name || `Entidad ${ec}`,
      sector: meta.sector === 0 ? 'financiero' : 'no_financiero',
      tipo: tipo === 0 ? 'familia' : 'empresa',
      monthsTC: new Float64Array(N_MONTHS),
      monthsIC: new Float64Array(N_MONTHS),
      monthsTR: new Int32Array(N_MONTHS),
      monthsIR: new Int32Array(N_MONTHS),
    };
    entData.set(key, e);
  }
  return e;
}

async function processFile(entities) {
  console.error(`Processing 24DSF.txt: ${INPUT_FILE}`);
  const stream = fs.createReadStream(INPUT_FILE, { encoding: 'latin1', highWaterMark: 1 << 22 });

  let buffer = '';
  let n = 0;
  const t0 = Date.now();

  await new Promise((resolve, reject) => {
    stream.on('data', (chunk) => {
      buffer += chunk;
      let idx;
      while ((idx = buffer.indexOf('\n')) !== -1) {
        let line = buffer.substring(0, idx);
        buffer = buffer.substring(idx + 1);
        if (line.endsWith('\r')) line = line.slice(0, -1);
        n++;
        if ((n % 10_000_000) === 0) {
          const elapsed = (Date.now() - t0) / 1000;
          console.error(`  ${n.toLocaleString()} lines (${elapsed.toFixed(1)}s, ${(n/elapsed).toFixed(0)} l/s)`);
        }

        if (line.length < BLOCK_START + BLOCK_SIZE) continue;

        // CUIT prefix at pos 7-8
        const cp0 = line.charCodeAt(7);
        const cp1 = line.charCodeAt(8);
        let tipo;
        // '2' is 50, '0'=48, '3'=51, '4'=52, '7'=55
        if (cp0 === 50 && (cp1 === 48 || cp1 === 51 || cp1 === 52 || cp1 === 55)) {
          tipo = 0; // familia
        } else if (cp0 === 51 && (cp1 === 48 || cp1 === 51 || cp1 === 52)) {
          tipo = 1; // empresa
        } else {
          continue;
        }

        const ec = line.substring(0, 5).trim().padStart(5, '0');
        const meta = entities[ec];
        const sector = meta ? meta.sector : classifyEntity(ec);
        const combo = sector * 2 + tipo;

        const entObj = getOrAddEntKey(entities, ec, tipo);

        for (let mi = 0; mi < N_MONTHS; mi++) {
          const off = BLOCK_START + mi * BLOCK_SIZE;
          if (off + 14 > line.length) break;

          // Situation: 2 chars
          const c0 = line[off];
          const c1 = line[off + 1];
          // Skip if both space or zero
          if ((c0 === ' ' || c0 === '0') && (c1 === ' ' || c1 === '0')) continue;

          const sit = (c0 + c1).replace(/\s/g, '');
          const amt = parseAmount(line.substring(off + 2, off + 14));

          const idx2 = combo * N_MONTHS + mi;
          aggTC[idx2] += amt;
          aggTR[idx2]++;

          entObj.monthsTC[mi] += amt;
          entObj.monthsTR[mi]++;

          if (IRREG_SITS.has(sit)) {
            aggIC[idx2] += amt;
            aggIR[idx2]++;
            entObj.monthsIC[mi] += amt;
            entObj.monthsIR[mi]++;
          }
        }
      }
    });
    stream.on('end', () => {
      // Process any leftover
      if (buffer.length > 0) {
        // ignore last partial
      }
      console.error(`Done: ${n.toLocaleString()} lines, ${entData.size} entity-tipo combinations`);
      resolve();
    });
    stream.on('error', reject);
  });
}

function buildLabels() {
  const mnames = ['', 'Ene','Feb','Mar','Abr','May','Jun','Jul','Ago','Sep','Oct','Nov','Dic'];
  const months = [];
  const labels = [];
  let y = LATEST_YEAR, m = LATEST_MONTH;
  for (let i = 0; i < N_MONTHS; i++) {
    months.push(`${y}-${String(m).padStart(2, '0')}`);
    labels.push(`${mnames[m]} ${String(y % 100).padStart(2, '0')}`);
    m--;
    if (m === 0) { m = 12; y--; }
  }
  return { months, labels };
}

function writeAggregateSeries(months, labels) {
  const keys = ['fam_fin', 'emp_fin', 'fam_nofin', 'emp_nofin'];
  // months/labels are in reverse chrono (block 0 = latest); we want oldest -> newest
  const out = {
    months: months.slice().reverse(),
    month_labels: labels.slice().reverse(),
    series: [],
  };
  // Build entries from oldest to newest (i = N_MONTHS-1 down to 0 means: we want to iterate from oldest to newest)
  for (let i = N_MONTHS - 1; i >= 0; i--) {
    const entry = {
      periodo: months[i],
      label: labels[i],
    };
    for (let c = 0; c < N_COMBOS; c++) {
      const idx = c * N_MONTHS + i;
      const tc = aggTC[idx], ic = aggIC[idx];
      const tr = aggTR[idx], ir = aggIR[idx];
      const pa = tc > 0 ? ic / tc * 100 : 0;
      const pq = tr > 0 ? ir / tr * 100 : 0;
      const k = keys[c];
      entry[`${k}_amt`] = round2(pa);
      entry[`${k}_qty`] = round2(pq);
      entry[`${k}_tc`] = round1(tc);
      entry[`${k}_ic`] = round1(ic);
      entry[`${k}_tr`] = tr;
      entry[`${k}_ir`] = ir;
    }
    out.series.push(entry);
  }

  fs.writeFileSync(path.join(OUTPUT_DIR, 'data_series.json'), JSON.stringify(out));
  console.error(`Saved data_series.json`);
}

function writeEntitySeries() {
  // Output JSON: { "00007_familia": {code, name, sector, tipo, months: [[tc,ic,tr,ir,pa,pq], ...]} }
  // Months ordered from oldest to newest
  const out = {};
  for (const [key, e] of entData.entries()) {
    // Check if has any data
    let hasData = false;
    for (let i = 0; i < N_MONTHS; i++) {
      if (e.monthsTR[i] > 0) { hasData = true; break; }
    }
    if (!hasData) continue;

    const monthsArr = [];
    for (let i = N_MONTHS - 1; i >= 0; i--) {
      const tc = e.monthsTC[i], ic = e.monthsIC[i];
      const tr = e.monthsTR[i], ir = e.monthsIR[i];
      const pa = tc > 0 ? ic / tc * 100 : 0;
      const pq = tr > 0 ? ir / tr * 100 : 0;
      monthsArr.push([round1(tc), round1(ic), tr, ir, round2(pa), round2(pq)]);
    }
    out[key] = {
      code: e.code,
      name: e.name,
      sector: e.sector,
      tipo: e.tipo,
      months: monthsArr,
    };
  }
  fs.writeFileSync(path.join(OUTPUT_DIR, 'data_entity_series.json'), JSON.stringify(out));
  console.error(`Saved data_entity_series.json (${Object.keys(out).length} entries)`);
}

function round1(x) { return Math.round(x * 10) / 10; }
function round2(x) { return Math.round(x * 100) / 100; }

(async () => {
  const t0 = Date.now();
  const entities = loadEntities();
  console.error(`Loaded ${Object.keys(entities).length} entities`);

  await processFile(entities);
  const { months, labels } = buildLabels();

  writeAggregateSeries(months, labels);
  writeEntitySeries();

  // Print last 6 months familias financiero summary
  console.error('\n=== Fam Fin (ultimos 6 meses) ===');
  for (let i = 5; i >= 0; i--) {
    const idx = 0 * N_MONTHS + i;
    const tc = aggTC[idx], ic = aggIC[idx];
    const p = tc > 0 ? ic/tc*100 : 0;
    console.error(`  ${labels[i]}: ${p.toFixed(2)}%`);
  }
  console.error(`\nTotal time: ${((Date.now() - t0)/1000).toFixed(1)}s`);
})();
